### DISCUSSION TITLE
Weird behavior with Obsidian installed from Flathub(Flatpak) and python

### DISCUSSION THREAD

**Discussion by zheneknew** [Taitava/obsidian-shellcommands#225] [Debugging and testing]
Hi!
My system is Fedora 35 Gnome 41.6
I had Obsidian installed from a default Gnome GUI apps store app named "Software"
![image](https://user-images.githubusercontent.com/69075015/169548243-170e4c15-9947-41b3-af6b-b32a515c5b69.png)
and it install Obsidian app from Flathub, not from dnf repos
![image](https://user-images.githubusercontent.com/69075015/169548536-f10540ed-b9f2-4f80-9a2e-b803fcf3238a.png)

At first I tried to create a shell-command to launch a python script from a specific virtual environment.
```bash
source /path/to/venv/bin/activate && python /path/to/script.py
```
and I got an exception
![image](https://user-images.githubusercontent.com/69075015/169543541-f61d87d3-2fc5-44ba-b8a0-74dd9accd321.png)
that means python virtual env was not activated
After trying to get different debug info I have decided to print python version from within script, I have created script.sh:
```bash
source /path/to/venv/bin/activate 
python  -c "import sys; print(sys.version, file=sys.stderr)"
```
created shell-command in your plugin
```bash
/path/to/script.sh
```
and got really weird output
![image](https://user-images.githubusercontent.com/69075015/169546455-be777a14-674c-4780-bbb7-e2549a4634ca.png)
It's weird because I haven't python 3.9 in my system, there is python3.10 out of the box in Fedora 35 and I didn't install other versions
The only python3.9 that i have find with `sudo find / -iname "python3.9*"` were 
```
/var/lib/flatpak/runtime/org.kde.Platform/x86_64/<some nested folders>/files/bin/python3.9
/var/lib/flatpak/runtime/org.freedesktop.Platform/x86_64/<some nested folders>/files/bin/python3.9
/var/lib/flatpak/runtime/org.gnome.Platform/x86_64/<some nested folders>/files/bin/python3.9
```
Then I downloaded Obsidian [AppImage version](https://github.com/obsidianmd/obsidian-releases/releases/download/v0.14.6/Obsidian-0.14.6.AppImage) and opened as vault the same folder, and this problem has disappeared, everything works fine, the virtualenv is activated, the python script works with the activated virtualenv, and therefore the script correctly imports all the modules.

I'm not sure what is the cause of this problem, may be it's because of Flatpak apps run in isolated env inside the OS, the fact that it runs   python3.9 is hinting at this.
I'm not sure that the way to fix this exists, so it may make sense to write in the "Known Issues" section that the Obsidian app installed from Flatpak limit the performance of your plugin

**Comment by Taitava** [comment]
Interesting. Thank you for reporting this! I will inspect this better later, just wanted to say I've seen your message. 🙂👍

**Comment by Taitava** [comment]
I finally came back to this. I'm not familiar with Flatpak, so I'm not going to try it out myself. But I added the mention to [README.md](https://github.com/Taitava/obsidian-shellcommands/blob/main/README.md) like you suggested.

Thank you again for reporting this! 🙂 If you or anyone else later comes up with more information that may help solve this problem, post here and I'll reconsider inspecting this. 👍

**Comment by a-matthew** [comment]
I can confirm the same issue even in the new FlatPak versions. Upon switching to AppImage the issue is gone, the entries in PATH are also properly detected. I tried adjusting the permissions through Flatseal with no effect. 

It's generally not limited to just this extension, as I ran into this issue in another plugin also - I assume that the Python binary in the Flatpak environment is neither a venv (which I also tried running by path) nor the global sytem binary (likely the runtime as suggested).

Offtopic, but related:
"Customizable Sidebar" mentioned in README is no longer used, but I noticed there is a quite popular plugin called "Buttons", which appears to have good synergy with this plugin and I managed to use it for custom Shell workflows. It works really well!

Thank you for your work, this plugin is great!

**Reply by Taitava** [reply]
> I can confirm the same issue even in the new FlatPak versions. Upon switching to AppImage the issue is gone, the entries in PATH are also properly detected. I tried adjusting the permissions through Flatseal with no effect.

Yea. Looks like I haven't updated this dicussion, but the SC plugin (since version `0.21.0` 2023-12-30) should nowadays warn users when it detects a Flatpak installation, and there's now also a [documentation page telling more about the Flatpak problem](https://publish.obsidian.md/shellcommands/Problems/Flatpak+installation) (maybe you already found it). But sometimes the warning has CSS related layout/visibility problems.

> Offtopic, but related: "Customizable Sidebar" mentioned in README is no longer used, but I noticed there is a quite popular plugin called "Buttons", which appears to have good synergy with this plugin and I managed to use it for custom Shell workflows. It works really well!

Thanks for letting me know. The "Benefits from other plugins" section in README.md has not been maintained in a long time. Well, actually the whole README.md is not actively maintained. I should go through it some day and remove/update obsolete stuff.

However, now that I looked at the [Buttons plugin's repo](https://github.com/shabegom/buttons), it looks like that's deprecated, too.


> Thank you for your work, this plugin is great!

Thank you for your kind feedback! ❤️

**Reply by Taitava** [reply]
I forgot to mention:

## If anyone comes up with a way to configure Flatpak so it allows executing certain user-whitelistable external programs, please let me know. 🙂 

I can then update the documentation of this plugin. I'm otherwise probably not using much time researching this problem.
